android/libraries/rib-base/src/main/java/com/badoo/ribs/core/Node.kt

```diff
@@ -183,30 +221,98 @@ open class Node<V : RibView>(
+    /**
+     * To be called only from the hosting environment (Activity, Fragment, etc.)
+     *
+     * For internal usage call onStopInternal() directly
+     */
+    fun onStop() {
+        externalLifecycleRegistry.handleLifecycleEvent(ON_STOP)
+        onStopInternal { it.onStop() }
+    }
 
-        router.onAttach(savedInstanceState?.getBundle(KEY_ROUTER))
-        interactor.onAttach(savedInstanceState?.getBundle(KEY_INTERACTOR))
+    /**
+     * To be called only from the hosting environment (Activity, Fragment, etc.)
+     *
+     * For internal usage call onResumeInternal() directly
+     */
+    fun onResume() {
+        externalLifecycleRegistry.handleLifecycleEvent(ON_RESUME)
+        onResumeInternal { it.onResume() }
     }
 
-    open fun onDetach() {
-        if (isViewAttached) {
-            RIBs.errorHandler.handleNonFatalError(
-                "View was not detached before node detach!",
-                RuntimeException("View was not detached before node detach!")
-            )
-            detachFromView()
+    /**
+     * To be called only from the hosting environment (Activity, Fragment, etc.)
+     *
+     * For internal usage call onPauseInternal() directly
+     */
+    fun onPause() {
+        externalLifecycleRegistry.handleLifecycleEvent(ON_PAUSE)
+        onPauseInternal { it.onPause() }
+    }
+
+    internal fun onStartInternal(callOnChildren: (Node<*>) -> Unit = { it.onStartInternal() }) {
+        // The lifecycle cannot go higher than that of the hosting environment (e.g. Activity)
+        updateToStateIfViewAttached(externalLifecycleRegistry.currentState)
+        children.forEach { callOnChildren.invoke(it) }
+    }
+
+    // TODO call this when overlay is removed
+    internal fun onResumeInternal(callOnChildren: (Node<*>) -> Unit = { it.onResumeInternal() }) {
+        // The lifecycle cannot go higher than that of the hosting environment (e.g. Activity)
+        updateToStateIfViewAttached(externalLifecycleRegistry.currentState)
+        children.forEach { callOnChildren.invoke(it) }
+    }
+
+    // TODO call this when overlay hides content AND current rib is not in dialog
+    internal fun onPauseInternal(callOnChildren: (Node<*>) -> Unit = { it.onPauseInternal() }) {
+        // The lifecycle cannot go higher than that of the hosting environment (e.g. Activity)
+        val targetState = if (externalLifecycleRegistry.currentState == CREATED) CREATED else STARTED
+        updateToStateIfViewAttached(targetState)
+        children.forEach { callOnChildren.invoke(it) }
+    }
+
+    internal fun onStopInternal(callOnChildren: (Node<*>) -> Unit = { it.onStopInternal() }) {
+        updateToStateIfViewAttached(CREATED)
+        children.forEach { callOnChildren.invoke(it) }
+    }
+
+    private fun updateToStateIfViewAttached(state: Lifecycle.State) {
+        if (isAttachedToView) {
+            ribLifecycleRegistry.markState(state)
+            view?.let { viewLifecycleRegistry.markState(state) }
```

Comment:
maybe `view!!`? otherwise, if we'll have an incorrect state with `null` view for some reason it will silently skip the lifecycle change

Reply:
There are RIBs without a view (purely business logic containers), so the above code is correct. It's the `isAttachedToView` name that is not expressive enough (but as of yet I don't have a better idea for it). It does not mean that *our* view is attached, it just means that this Node and its parents are aware of the screen.

Reply:
What do you think about checking like this? I think it will be safer to immediately fail if we know that it should not happen

```
if (isAttachedToView) {
    ...
    if (viewFactory != null) {
        view!!.let { ... }
    }
}
```

Also, maybe you can create a property like this:
```
private val isViewless get() = viewFactory == null
```

Reply:
Done